### PR TITLE
[clang-tidy] fix match for binaryOperator in ExprMutationAnalyzer for misc-const-correctness

### LINKED ISSUE #57297: `misc-const-correctness` false positive with variable in template
```cpp
#include <sstream>
#include <string>

template<class T> static std::string toString(T t)
{
    std::ostringstream ostr; // warning
    ostr << t;
    return ostr.str();
}

std::string toStringF(int i)
{
    std::ostringstream ostr; // no warning
    ostr << i;
    return ostr.str();
}

static void f()
{
    toString(1);
    toStringF(1);
}
```

```console
<source>:6:5: warning: variable 'ostr' of type 'std::ostringstream' (aka 'basic_ostringstream<char>') can be declared 'const' [misc-const-correctness]
    std::ostringstream ostr; // warning
    ^
                       const
```

https://godbolt.org/z/8n1K8dfEx

Fixing the warning results in a compiler error:
```console
error: invalid operands to binary expression ('const std::ostringstream' (aka 'const basic_ostringstream<char>') and 'long long')
    ostr << t;
    ~~~~ ^  ~
```

### PR DESCRIPTION
**Pull request by 5chmidti**
The `ExprMutationAnalyzer`s matcher of `binaryOperator`s
that contained the variable expr, were previously narrowing the
variable to be type dependent, when the `binaryOperator` should
have been narrowed as dependent.
The variable we are trying to find mutations for does
not need to be the dependent type, the other operand of
the `binaryOperator` could be dependent.

Fixes #57297

### CODE REVIEW

**Comment by 5chmidti** [comment]
CC @JonasToth for your work on `ExprMutationAnalyzer`

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -683,6 +683,9 @@ Bug Fixes to AST Handling
 - Fixed a bug where RecursiveASTVisitor fails to visit the
   initializer of a bitfield.
   `Issue 64916 <https://github.com/llvm/llvm-project/issues/64916>`_
+- Fixed a bug where ``ExprMutationAnalyzer`` did not find a potential mutation
```
Comment (PiotrZSL) [inline clang/docs/ReleaseNotes.rst:686]: I think this should be under static analyzer changes or not included at all.

Comment (5chmidti) [reply clang/docs/ReleaseNotes.rst]: Done

**Comment by 5chmidti** [comment]
The top-level Linux and Windows tests work, but inside the `Trigger Build`/`clang-ci` run, the windows test step fails. I'm not sure what the difference is between these two Windows tests.

**Comment by jcsxky** [comment]
> The top-level Linux and Windows tests work, but inside the `Trigger Build`/`clang-ci` run, the windows test step fails. I'm not sure what the difference is between these two Windows tests.

This is interesting! Generally speaking, AST of the code in Windows and Linux platform is identical with clang compiled. Probably need debuging in Windows.

**Comment by 5chmidti** [comment]
I found the issue, but I have not looked for a solution.
See https://godbolt.org/z/or4EnMfWj and #71203.
When using clang with `-target` to specify `x86_64-pc-win32` and `x86_64-linux`, the body for the function `f` is `NULL` in the AST with the windows target. This only happens for `-target x86_64-windows` pre `c++20`.

Although the tested code does not directly depend on being `c++20` (it is valid in `c++11`), maybe I constrain the tests in this pr to be in `c++20` mode until the issue is fixed? What do you think?

**Comment by jcsxky** [comment]
how about adding a function call like `f<int>()`? will this can prevent clang to emit empty function body?

**Comment by 5chmidti** [comment]
> how about adding a function call like `f<int>()`? will this can prevent clang to emit empty function body?

Yep, thanks. I went this route, but I had to define a conversion operator to compile and keep the `<<` a `BinaryOperator`. The CI now passes

**Comment by PiotrZSL** [comment]
In tests we usually use `-fno-delayed-template-parsing` to overcome this issue with templates and windows.

**Comment by 5chmidti** [comment]
> In tests we usually use `-fno-delayed-template-parsing` to overcome this issue with templates and windows.

Done. I changed the test in `ExprMutationAnalyzerTest.cpp` to use the flag, `const-correctness-templates.cpp` already had it, and I changed the tested code back to what it was before (-formatting).

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -308,6 +308,11 @@ Changes in existing checks
   <clang-tidy/checks/misc/const-correctness>` check to avoid false positive when
   using pointer to member function.
 
+- Improved :doc:`misc-const-correctness
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:311]: there is already entry for this check, just merge changes.

**Review by PiotrZSL — APPROVED** [review]

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 2 above)
Comment (5chmidti) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done. I have also reworded the release note, I think it reads better now.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -307,11 +307,9 @@ Changes in existing checks
 - Improved :doc:`misc-const-correctness
   <clang-tidy/checks/misc/const-correctness>` check to avoid false positive when
   using pointer to member function.
-
-- Improved :doc:`misc-const-correctness
-  <clang-tidy/checks/misc/const-correctness>` check to not warn on uses in
-  type-dependent binary operators, when the variable that is being
-  looked at, is not the dependent operand.
+  Additionally, the check no longer emits a diagnostic when
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:310]: Looks fine, but wrap this after "to member function."

Comment (5chmidti) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Like this?

**Pull request merged**